### DISCUSSION TITLE
true_divide error

### DISCUSSION THREAD

**Discussion by kasia129** [capytaine/capytaine#130] [Q&A]
Hi, I am getting the following error when running Capytaine:

```
...\capytaine\meshes\properties.py:35: RuntimeWarning: invalid value encountered in true_divide
  faces_normals[triangles_id] = triangles_normals / np.array(([triangles_areas, ] * 3)).T
```

What may be the reason for such error? I guess for some reason the `triangles_areas` is computed to be zero, hence it's not possible to perfrom the division, but I can't see why that is the case.

As a background, I'm running Capytaine within an optimization framework on multiple geometries, and this error only pops out for some of the meshes. I'm creating my meshes with:

```
buoy = cpt.FloatingBody(
        cpt.AxialSymmetricMesh.from_profile(profile=shape,...)
```

I would be grateful for any hints.

**Comment by mancellin** [comment]
Yes, this warning is usually due to a division by zero.
It might be possible to end up with a face of area zero, for instance if there are two identical points in the `shape` of the axisymmetric mesh.

**Reply by kasia129** [reply]
Thanks. My `shape` was defined as follows: 

```
[[  0.     0.   -80.4 ]
 [  1.     0.   -80.4 ]
 [  2.     0.   -80.4 ]
 [  3.     0.   -80.4 ]
 [  3.0000000084834726     0.   -77.05]
 [  3.0000001324596766     0.   -73.7 ]
  ... 
 [  3.0000000000103     0.    -3.35]
 [  3.     0.     0.  ]]
```
Cutting the 'tail' and only leaving 4 decimal places on all coordinates solved the problem. I wonder where the problem with dealing with large precision stems from and whether the code could be fool-proved?

**Reply by mancellin** [reply]
I cannot reproduce the warning message. Do you have a minimal reproducible example that I could use?

<details>
<summary>My attempt to reproduce</summary>

```python
import numpy as np
import capytaine as cpt

shape = np.array([
        [  0.,     0.,   -80.4 ],
        [  1.,     0.,   -80.4 ],
        [  2.,     0.,   -80.4 ],
        [  3.,     0.,   -80.4 ],
        [  3.0000000084834726,     0.,   -77.05],
        [  3.0000001324596766,     0.,   -73.7 ],
        [  3.0000000000103,     0.,    -3.35],
        [  3.,     0.,     0.  ],
        ])

mesh = cpt.AxialSymmetricMesh.from_profile(profile=shape)
mesh.faces_normals
```

</details>

**Reply by kasia129** [reply]
I am trying to reproduce the problem again but I seem not to be getting this error now. I am assuming this error might have had to do with the clipping operation: I used the buoy.keep_immersed_part() function to cut the mesh at the waterline. Perhaps this left the first set of panels at just below the waterline with a marginally small area? I will try to reproduce this error again once I get a bit more time, but it seems that this is not an issue that users will encounter too often. Thanks for your help anyway!
